### DISCUSSION TITLE
Sets of labels for a node

### DISCUSSION THREAD

**Discussion by GrimPixel** [IndiePython/nodezator#73] [Feature Request]
There is a possibility of optimisation: aggregating formulae in the same structure.
For example, the node `（Ａ＋Ｂ）Ｃ／２`, or if following the Python 2's naming rule `opApBcpCs2`, which represents $X = (A+B) C/2$:
* area of a trapezoid: $A = (a+b) h/2$
* displacement in an uniformly accelerated linear motion $x = (v_0+v_t) t/2$

I hope to find a way to circumvent that naming restriction, so the node names are easily readable.

They can be aggregated to one node with two sets of labels in different languages:
```
labels = [
    ['area of a trapezoid', 'upper base', 'lower base', 'height'], # eng-Latn-US
    ['梯形面积', '上底', '下底', '高'], # cmn-Hans-CN
    ['A,S', 'a', 'b', 'h'], # zxx-Zsym-ZZ

    ['displacement of a uniformly accelerated linear motion', 'start velocity', 'end velocity', 'time'], # eng-Latn-US
    ['匀加速直线运动位移', '初速度', '末速度', '时间'],  # cmn-Hans-CN
    ['x,s', 'v_0', 'v_t', 't'], # zxx-Zsym-ZZ
]
```

Labels can be searched as well.

It saves plenty of work if implemented.

![](https://upload.wikimedia.org/wikipedia/commons/thumb/2/21/Basic_physics_formula_triangles.svg/512px-Basic_physics_formula_triangles.svg.png)

**Comment by GrimPixel** [comment]
There is an alternative solution: allow functions to have aliases. They will look like this:
`X_area_of_a_trapezoid__A_upper_base__B_lower_base__C_height`
`X_梯形面积__A_上底__B_下底__C_高`
This still requires lifting the restriction on function names and is less intuitive.

**Reply by KennedyRichard** [reply]
Yes, I just commented on this on the other discussion. There are restrictions on the names that Python imposes and there's nothing we can do about them.

My recommendation is to adopt simpler names like `area_of_trapezoid`, `area_of_trapezoid_2`, `area_of_trapezoid_3` or something similar and let the user search for the simpler terms on the search bar when it is implemented.

For instance, searching for `trapezoid` should show the available functions containing that name, and the user would try the different available nodes until a satisfactory one is found. The search could also show an image of the node's body, so the user could easily see the node they want to instantiate (it's parameters and outputs).

**Reply by GrimPixel** [reply]
Uncertainties are brought in if I number them simply 1, 2, 3. The user would have to hover the mouse over them one by one to find the one needed. It requires some rote learning.

So, the ideal solution consists of these traits:
1. There are no duplicated functions for a same calculation process.
2. Names for inputs and outputs are displayed on the node.
3. Node identifiers are intuitive.
4. Node identifiers are language-neutral.

I can only guarantee the first for the time being.

Another approach comes into my mind: keep templates for a same calculation process and nodes are duplicated from them with names of inputs and outputs changed only.

**Reply by KennedyRichard** [reply]
Fair enough. Although simpler names are best, just numbering the names would not be clear enough. It may work in some cases but for your purposes it would not fit. It was hasty on my part.

The approach you mentioned at the end of your reply would be great, you wanted something like that from the beginning. But I'm not sure Python has something like that. That is, changing a function by inverting the operations performed like we do with equations to calculate different variables. Remember that Nodezator aims to work as close to Python as possible.

I'll keep all your ideas and suggestions in mind though. I'm actually interested in this problem because finding a solution to it would be really useful for handling equations not only in Nodezator but in Python in general.

I even did some research on my end but didn't come up with a solution yet. I found a third-party library (called sympy) that is awesome, but even though it handles math notation it doesn't dynamically generates callables or solutions like we need. With time I think we'll be able to find a suitable solution.

**Reply by GrimPixel** [reply]
It is great to know that you are hooked, too. Let me summerise all proposals.

# each file contains one unique equation
## node names are the output and inputs

folders for files:
* `area_of_a_trapezoid___upper_base__lower_base__height`
* `displacement_of_a_uniformly_accelerated_linear_motion___start_velocity__end_velocity__time`

file contents:
```
def area_of_a_trapezoid___upper_base__lower_base__height(a:float=0.0, b:float=0.0, h:float=0.0) -> [{"name": "A", "type": float}]
	return (a+b)*h/2
```
```
def displacement_of_a_uniformly_accelerated_linear_motion___start_velocity__end_velocity__time(v_0:float=0.0, v_t:float=0.0, t:float=0.0) -> [{"name": "x", "type": float}]
	return (v_0+v_t)*t/2
```

## node names are outputs with numbers

folders for files:
* `area_of_a_trapezoid_1`
* `displacement_of_a_uniformly_accelerated_linear_motion_1`

file contents:
```
def area_of_a_trapezoid_1(a:float=0.0, b:float=0.0, h:float=0.0) -> [{"name": "A", "type": float}]
	return (a+b)*h/2
```
```
def displacement_of_a_uniformly_accelerated_linear_motion_1(v_0:float=0.0, v_t:float=0.0, t:float=0.0) -> [{"name": "x", "type": float}]
	return (v_0+v_t)*t/2
```

## node names are the output and inputs, represented by symbols

folders for files:
* `X_area_of_a_trapezoid___A_upper_base__B_lower_base__C_height`
* `X_displacement_of_a_uniformly_accelerated_linear_motion___A_start_velocity__B_end_velocity__C_time`

file contents:
```
def area_of_a_trapezoid___upper_base__lower_base__height(A:float=0.0, B:float=0.0, C:float=0.0) -> [{"name": "X", "type": float}]
	return (A+B)*C/2
```
```
def displacement_of_a_uniformly_accelerated_linear_motion___start_velocity__end_velocity__time(A:float=0.0, B:float=0.0, C:float=0.0) -> [{"name": "X", "type": float}]
	return (A+B)*C/2
```

# each file contains one unique equation calculation
## node names are fullwidth characters of parameters (invalid)

folders for files:
* `（Ａ＋Ｂ）Ｃ／２`

file contents:
```
def （Ａ＋Ｂ）Ｃ／２(A:float=0.0, B:float=0.0, C:float=0.0) -> [{"name": "X", "type": float}]
	return (A+B)*C/2
```

## node names are the calculation with text with escape characters (here XML/HTML)

folders for files:
* `lpar_A_plus_B_rpar_C_sol_2`

file contents:
```
def lpar_A_plus_B_rpar_C_sol_2(A:float=0.0, B:float=0.0, C:float=0.0) -> [{"name": "X", "type": float}]
	return (A+B)*C/2
```

## node names are texts in a natural or constructed language (here: modified Lojban)

folders for files:
* `vei_ay0_suhi_by0_veho_cy0_fihu_re`

file contents:
```
def vei_ay0_suhi_by0_veho_cy0_fihu_re(A:float=0.0, B:float=0.0, C:float=0.0) -> [{"name": "X", "type": float}]
	return (A+B)*C/2
```

## node names are the hash value with an underscore (here: MD5)

folders for files:
* `_4fe252e3443e2f8e71b0a96ba938959d`

file contents:
```
def _4fe252e3443e2f8e71b0a96ba938959d(A:float=0.0, B:float=0.0, C:float=0.0) -> [{"name": "X", "type": float}]
	return (A+B)*C/2
```

# each derived file contains one unique equation, imported from a base file

folders for files:
* `_4fe252e3443e2f8e71b0a96ba938959d`
* `area_of_a_trapezoid___upper_base__lower_base__height`
*  `displacement_of_a_uniformly_accelerated_linear_motion___start_velocity__end_velocity__time`

file contents:
```
def _4fe252e3443e2f8e71b0a96ba938959d(A:float=0.0, B:float=0.0, C:float=0.0) -> [{"name": "X", "type": float}]
	return (A+B)*C/2
```
```
import _4fe252e3443e2f8e71b0a96ba938959d._main_._4fe252e3443e2f8e71b0a96ba938959d
def area_of_a_trapezoid___upper_base__lower_base__height(a:float=0.0, b:float=0.0, h:float=0.0, b:float=0.0) -> [{"name": "A", "type": float}]
	return _4fe252e3443e2f8e71b0a96ba938959d(a, b, h)
```
```
import _4fe252e3443e2f8e71b0a96ba938959d._main_._4fe252e3443e2f8e71b0a96ba938959d
def displacement_of_a_uniformly_accelerated_linear_motion___start_velocity__end_velocity__time(v_0:float=0.0, v_t:float=0.0, t:float=0.0) -> [{"name": "x", "type": float}]
	return _4fe252e3443e2f8e71b0a96ba938959d(v_0, v_t, t)
```

# each file contains forms of similar formulae
folders for files:
* `area_of_a_trapezoid__upper_base__lower_base__height`
* `displacement_of_a_uniformly_accelerated_linear_motion__start_velocity__end_velocity__time`

file contents:
```
def area_of_a_trapezoid__upper_base__lower_base__height(A:float=0.0, a:float=0.0, b:float=0.0, h:float=0.0)
	if A is null and null not in [a, b, h]:
		return (a+b)*h/2
	elif a is null and null not in [A, b, h]:
		return 2*A/h-b
	elif b is null and null not in [A, a, h]:
		return 2*A/h-a
	elif h is null and null not in [A, a, b]:
		return 2*A/(a+b)
```

```
def displacement_of_a_uniformly_accelerated_linear_motion__start_velocity__end_velocity__time(x:float=0.0, v_0:float=0.0, v_t:float=0.0, t:float=0.0)
	if x is null and null not in [v_0, v_t, t]:
		return (v_0+v_t)*t/2
	elif v_0 is null and null not in [x, v_t, t]:
		return 2*x/t-v_t
	elif v_t is null and null not in [x, v_0, t]:
		return 2*x/t-v_0
	elif t is null and null not in [x, v_0, v_t]:
		return 2*x/(v_0+v_t)
```

# each file contains forms of similar formulae calculation, imported from a base file, with a name of hash value (here: MD5)
folders for files:
* `_7e0bb96c9462f6a2e4d921dc491fb0b7`
* `area_of_a_trapezoid__upper_base__lower_base__height`
* `displacement_of_a_uniformly_accelerated_linear_motion__start_velocity__end_velocity__time`

file contents:

```
def _7e0bb96c9462f6a2e4d921dc491fb0b7(A:float=0.0, B:float=0.0, C:float=0.0, D:float=0.0) -> [{"name": "X", "type": float}]
    if A is null and null not in [B, C, D]:
		return (B+C)*D/2
	elif B is null and null not in [A, C, D]:
		return 2*A/D-C
	elif C is null and null not in [A, B, D]:
		return 2*A/D-B
	elif D is null and null not in [A, B, C]:
		return 2*A/(B+C)
```

```
import _7e0bb96c9462f6a2e4d921dc491fb0b7._main_._7e0bb96c9462f6a2e4d921dc491fb0b7
def area_of_a_trapezoid__upper_base__lower_base__height(A:float=0.0, a:float=0.0, b:float=0.0, h:float=0.0)
    return _7e0bb96c9462f6a2e4d921dc491fb0b7(A, a, b, h)
```
```
import _7e0bb96c9462f6a2e4d921dc491fb0b7._main_._7e0bb96c9462f6a2e4d921dc491fb0b7
def displacement_of_a_uniformly_accelerated_linear_motion__start_velocity__end_velocity__time(x:float=0.0, v_0:float=0.0, v_t:float=0.0, t:float=0.0)
    return _7e0bb96c9462f6a2e4d921dc491fb0b7(x, v_0, v_t, t)
```

**Reply by GrimPixel** [reply]
There are dimensions:
# naming 
1. output with numbering
2. output and input with text in a natural or constructed language
3. ~~input with fullwidth characters~~
4. input with text with escape characters
5. input with a hash value with an underscore
6. the other side of 0 of a formula with text with escape characters
7. the other side of 0 of a formula with a hash value with an underscore

# meaning of a node
1. a unique formula
3. a unique calculation for similar formulae
2. all forms of a unique formula
4. all forms of a unique calculation for similar formulae

# dependency
1. files are independent from each other
2. some files are the base and others depend on them

# labeling for node names, output and input names
1. not applied
2. applied


I tend to choose the combination that represents the nature of calculation to my understanding, so I prefer the combination of “input with text with escape characters (XML/HTML)”, “a unique calculation for similar formulae”, “files are independent from each other”, “applied”.

**Reply by GrimPixel** [reply]
It is working now.
![result](https://github.com/IndiePython/nodezator/assets/39795763/365b9ab9-c494-4efb-b253-35b878c0954d)

The font doesn't include tab and non-Japanese Han characters.
![source](https://github.com/IndiePython/nodezator/assets/39795763/9c80ad9b-bb47-4c63-ad59-63164a5260fc)

I think it serves the educational purpose in a flashcard-way: check if your memory of a formula is correct.

Update: Node names are Python + XML/HTML now. What you see is what you get.

Heron's formula

![formula](https://github.com/IndiePython/nodezator/assets/39795763/f21becaf-5892-4fa5-b3f1-d949d98a0a0e)

**Reply by KennedyRichard** [reply]
Thank you for the detailed feedback. I appreciate it a lot. This shows a lot of dedication.

When it is the time to implement some of the agreed features or study them better to determine what should be implemented I'll certainly revisit this content and take it into consideration, as well as link other people/discussions to it so others can give feedback as well.

**Reply by GrimPixel** [reply]
I am thinking of moving the labels from the `.py` file to a `.yaml` or a `.tsv` file instead, so the `.py` files are clean as they should be. This saves the memory when loading nodes, too. In addition, it saves the efforts of typing quotation marks and commas.

Since it's your wish to avoid external packages, I will do with TSV. The first line of the file is the text of input/output on the node, the corresponding fields are the labels.

`pi_ast_A_ast_B.tsv`:
```
X	A	B	
椭圆面积	半长轴长	半短轴长	cmn-Hans-CN
area of an ellipse	semi-major axis	semi-minor axis	eng-Latn-US
A,S	a	b	zxx-Zsym-ZZ
```

**Comment by KennedyRichard** [comment]
Also, although Python recommends ascii/English characters for names in Python, it is up to the users whether they decide using them or not. As long as the names are valid Python identifiers, it doesn't matter if they use non-ascii characters.

Since Python itself doesn't restrict the characters used, I see no problem implementing this possibility in Nodezator as well (instead of just letting people write non-english characters in widgets and comments, like I said before on another discussion). However, this is something that will take time until I implement, because Nodezator has other priority features to be implemented before. But I'll keep all of this in mind and I do think people must have the option to choose whether they want to write their code in English or their mother tongue.

**Comment by GrimPixel** [comment]
I will try another approach: build a database and generate node files from it.

**Reply by KennedyRichard** [reply]
Interesting. Although this is a workaround, it might be worth considering what you are trying to accomplish.

I've been working on Nodezator's priority features, but I didn't forget about your suggestions/requests. As I said, I'll consider them for implementation once other basic missing features are implemented, although it will take time.

The automated tests that I'm implementing right now are an important step towards this, since it will allow me to develop faster and much safer (with less chance of breaking existing stuff).

Have a nice weekend.

Peace.
